ulib/FStar.List.Pure.Base.fst

```diff
@@ -0,0 +1,55 @@
+module FStar.List.Pure.Base
+
+open FStar.List.Tot.Base
+
+(** Functions on list with a pure specification *)
+
+(** [map2] takes a pair of list of the same length [x1; ...; xn] [y1; ... ; yn]
+ and return the list [f x1 y1 ; ... ; f xn yn ] *)
+val map2 (#a1 #a2 #b: Type)
+  (f: a1 -> a2 -> b)
+  (l1:list a1)
+  (l2:list a2)
+  : Pure (list b)
+    (requires (length l1 == length l2))
+    (ensures (fun _ -> True))
+    (decreases l1)
+let rec map2 #a1 #a2 #b f l1 l2 =
+  match l1, l2 with
+  | [], [] -> []
+  | x1::xs1, x2::xs2 -> f x1 x2 :: map2 f xs1 xs2
+
+(** [map2] takes three lists of the same length [x1; ...; xn]
+    [y1; ... ; yn] [z1; ... ; zn] and return the list
+    [f x1 y1 z1 ; ... ; f xn yn zn ] *)
+val map3 (#a1 #a2 #a3 #b: Type)
+  (f: a1 -> a2 -> a3 -> b)
+  (l1:list a1)
+  (l2:list a2)
+  (l3:list a3)
+  : Pure (list b)
+    (requires (let n = length l1 in
+      (n == length l2 /\
+        n == length l3)))
```

Comment:
I was thinking of doing the 4 other maps (`map4`, ...) and I wanted to be consistent on th eindentation. Finally I got too lazy.